# Model Selection Criteria (Calculating AIC, BIC and MDL )
    
Example : Linear Regression

- Using a test problem and fiting a linear regression model, then evaluating the model using the AIC and BIC metrics.
- The problem will have two input variables and require the prediction of a target numerical value.

In [2]:
from sklearn.datasets import make_regression

# generate dataset
X, y = make_regression(n_samples=100, n_features=2, noise=0.1)

In [3]:
from sklearn.linear_model import LinearRegression

# define and fit the model on all data
model = LinearRegression()
model.fit(X, y)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

Once fit, The number of  parameters in the model is calculated.
- Expected : 3(two coefficients and one intercept).

In [4]:
# number of parameters
num_params = len(model.coef_) + 1
print('Number of parameters: %d' % (num_params))

Number of parameters: 3


The likelihood function for a linear regression model can be shown to be identical to the least squares function; therefore, we can estimate the maximum likelihood of the model via the mean squared error metric.

1. the model can be used to estimate an outcome for each example in the training dataset.
2. the mean_squared_error() scikit-learn function can be used to calculate the mean squared error for the model.

In [9]:
from sklearn.metrics import mean_squared_error

# predict the training set
yhat = model.predict(X)
# calculate the error
mse = mean_squared_error(y, yhat)
print('MSE: %.3f' % mse)

MSE: 0.008


**Reporting**
- Number of Parameters: 3
- MSE value : 0.008

The specific MSE value may vary running each time, due to the stochastic nature of the learning algorithm.

## AIC

The AIC calculation for an ordinary least squares linear regression model can be calculated as follows (taken from “A New Look At The Statistical Identification Model“,  1974.):

$AIC = n * LL + 2 * k$

Where $n$ is the number of examples in the training dataset, $LL$ is the log-likelihood for the model using the natural logarithm (e.g. the log of the MSE), and $k$ is the number of parameters in the model.

#### Defining AIC function

- Function : calculate_aic() 
- Arguments : n(no. of examples), the raw mean squared error (mse), and k(no. of parameters).

In [10]:
from math import log

# calculate aic for regression
def calculate_aic(n, mse, num_params):
    aic = n * log(mse) + 2 * num_params
    return aic

aic = calculate_aic(len(y), mse, num_params)
print('AIC: %.3f' % aic)

AIC: -477.529


**This value can be minimized in order to choose better models.**

## BIC

*The same example can be explored to calculate BIC* 

The BIC calculation for an ordinary least squares linear regression model can be calculated as follows (taken from [wikipedia-BIC-GaussianSpecialCase](https://en.wikipedia.org/wiki/Bayesian_information_criterion#Gaussian_special_case)):

$BIC = n * LL + k * log(n)$

Where $n$ is the number of examples in the training dataset, $LL$ is the log-likelihood for the model using the natural logarithm (e.g. log of the mean squared error), and $k$ is the number of parameters in the model, and $log()$ is the natural logarithm.


#### Defining BIC function

- Function : calculate_bic() 
- Arguments : n(no. of examples), the raw mean squared error (mse), and k(no. of parameters).

In [11]:
# calculate bic for regression
def calculate_bic(n, mse, num_params):
    bic = n * log(mse) + num_params * log(n)
    return bic

bic = calculate_bic(len(y), mse, num_params)
print('BIC: %.3f' % bic)

BIC: -469.713


*Again :* the results may vary given the stochastic nature of the learning algorithm.

In this case, the BIC is reported to be a value of about -469.713, which is very close to the AIC value of -477.529. Again, this value can be minimized in order to choose better models.

---

## MDL

The Minimum Description Length (MDL) principle recommends choosing the hypothesis that minimizes the sum of following two description lengths.

$MDL = L(h) + L(D | h)$

Where $h$ is the model, $D$ is the predictions made by the model, $L(h)$ is the number of bits required to represent the model, and $L(D | h)$ is the number of bits required to represent the predictions from the model on the training dataset

> *minimum number of bits, or the minimum of the sum of the number of bits required to represent the data and the model.(i.e. The model with the lowest MDL) is selected*


Since the score needs to be minimized, the number of bits required to encode $(D | h)$ and the number of bits required to encode $(h)$ can be calculated as the negative log-likelihood; the negative log-likelihood of the model parameters (theta) and the negative log-likelihood of the target values (y) given the input values (X) and the model parameters (theta).

$MDL = -log(P(\theta)) – log(P(y | X, \theta))$

#### Defining MDL function

- Function : calculate_mdl() 
- Arguments : 

In [14]:
# Steps 

# 1. calculating negative log likelihood of model parameter

# 2. calculating negative log likelihood of y given X and theta

## https://github.com/maxpumperla/entropy-mdlp/blob/master/entropymdlp/mdlp.py(- wanna try)


***